File: NeuroSim.Engine/Genomes/GraphGenome.cs
// AUTH:DEVNEUROSIM:7A3F9E2B | Legacy/NeuroSim.Engine/Genomes/GraphGenome.cs
using System.Text;

namespace NeuroSim.Engine.Genomes;

/// <summary>Graph genome — adjacency matrix + edge weights. Useful for network design.</summary>
public sealed class GraphGenome : Genome
{
    private readonly bool[,] _adj;
    private readonly double[,] _weights;

    public int NodeCount { get; }
    public bool Directed { get; }

    public GraphGenome(int nodeCount, bool directed = false)
    {
        if (nodeCount <= 0) throw new ArgumentOutOfRangeException(nameof(nodeCount));
        NodeCount = nodeCount;
        Directed = directed;
        _adj = new bool[nodeCount, nodeCount];
        _weights = new double[nodeCount, nodeCount];
    }

    private GraphGenome(GraphGenome src)
    {
        NodeCount = src.NodeCount;
        Directed = src.Directed;
        _adj = (bool[,])src._adj.Clone();
        _weights = (double[,])src._weights.Clone();
    }

    public override GenomeType Type => GenomeType.Graph;

    // Length = number of possible edges
    public override int Length => Directed
        ? NodeCount * (NodeCount - 1)
        : NodeCount * (NodeCount - 1) / 2;

    public bool HasEdge(int i, int j) => _adj[i, j];
    public double EdgeWeight(int i, int j) => _weights[i, j];

    public void SetEdge(int i, int j, bool present, double weight = 1.0)
    {
        _adj[i, j] = present;
        _weights[i, j] = weight;
        if (!Directed)
        {
            _adj[j, i] = present;
            _weights[j, i] = weight;
        }
    }

    public int EdgeCount()
    {
        int count = 0;
        for (int i = 0; i < NodeCount; i++)
            for (int j = Directed ? 0 : i + 1; j < NodeCount; j++)
                if (i != j && _adj[i, j]) count++;
        return count;
    }

    public IEnumerable<(int from, int to, double weight)> Edges()
    {
        for (int i = 0; i < NodeCount; i++)
            for (int j = Directed ? 0 : i + 1; j < NodeCount; j++)
                if (i != j && _adj[i, j])
                    yield return (i, j, _weights[i, j]);
    }

    public List<int> Neighbors(int node)
    {
        var n = new List<int>();
        for (int j = 0; j < NodeCount; j++)
            if (j != node && _adj[node, j]) n.Add(j);
        return n;
    }

    public override void Randomize(Random rng)
    {
        for (int i = 0; i < NodeCount; i++)
            for (int j = Directed ? 0 : i + 1; j < NodeCount; j++)
            {
                if (i == j) continue;
                bool present = rng.NextDouble() < 0.3;
                double w = rng.NextDouble();
                SetEdge(i, j, present, w);
            }
    }

    public override string Serialize()
    {
        var sb = new StringBuilder();
        sb.Append($"Graph(N={NodeCount}, E={EdgeCount()}): ");
        foreach (var (f, t, w) in Edges())
            sb.Append($"{f}->{t}({w:F2}) ");
        return sb.ToString().TrimEnd();
    }

    public override Genome Clone() => new GraphGenome(this);

    /// <summary>Get adjacency matrix as flat bool array for crossover.</summary>
    public bool[] FlattenAdjacency()
    {
        var flat = new List<bool>();
        for (int i = 0; i < NodeCount; i++)
            for (int j = Directed ? 0 : i + 1; j < NodeCount; j++)
                if (i != j) flat.Add(_adj[i, j]);
        return flat.ToArray();
    }

    /// <summary>Restore adjacency from flat bool array.</summary>
    public void UnflattenAdjacency(bool[] flat, Random rng)
    {
        int k = 0;
        for (int i = 0; i < NodeCount && k < flat.Length; i++)
            for (int j = Directed ? 0 : i + 1; j < NodeCount && k < flat.Length; j++)
            {
                if (i == j) continue;
                double w = HasEdge(i, j) ? EdgeWeight(i, j) : rng.NextDouble();
                SetEdge(i, j, flat[k++], w);
            }
    }
}

File: NeuroSim.Core/src/genome/GraphGenome.cpp
#include "../../include/genome/GraphGenome.h"
#include <sstream>
#include <queue>
#include <unordered_set>

namespace NeuroSim {

std::mt19937& GraphGenome::Rng() {
    thread_local static std::mt19937 rng{ std::random_device{}() };
    return rng;
}

GraphGenome::GraphGenome(int nodeCount, bool directed)
    : nodeCount_(nodeCount), directed_(directed),
      adj_(nodeCount * nodeCount, false),
      weights_(nodeCount * nodeCount, 0.0) {}

GraphGenome::GraphGenome(const GraphGenome& o)
    : nodeCount_(o.nodeCount_), directed_(o.directed_),
      adj_(o.adj_), weights_(o.weights_) {}

std::unique_ptr<IGenome> GraphGenome::Clone() const {
    return std::make_unique<GraphGenome>(*this);
}

bool GraphGenome::HasEdge(int i, int j) const { return adj_[Idx(i,j)]; }
double GraphGenome::EdgeWeight(int i, int j) const { return weights_[Idx(i,j)]; }

void GraphGenome::SetEdge(int i, int j, bool present, double weight) {
    adj_[Idx(i,j)] = present;
    weights_[Idx(i,j)] = weight;
    if (!directed_) {
        adj_[Idx(j,i)] = present;
        weights_[Idx(j,i)] = weight;
    }
}

int GraphGenome::EdgeCount() const {
    int count = 0;
    for (int i = 0; i < nodeCount_; ++i)
        for (int j = (directed_ ? 0 : i+1); j < nodeCount_; ++j)
            if (i != j && adj_[Idx(i,j)]) ++count;
    return count;
}

void GraphGenome::Randomize() {
    std::uniform_real_distribution<> wDist(0, 1);
    std::bernoulli_distribution eDist(0.3);
    for (int i = 0; i < nodeCount_; ++i)
        for (int j = (directed_ ? 0 : i+1); j < nodeCount_; ++j)
            if (i != j) SetEdge(i, j, eDist(Rng()), wDist(Rng()));
}

std::string GraphGenome::Serialize() const {
    std::ostringstream oss;
    oss << "Graph(N=" << nodeCount_ << ", E=" << EdgeCount() << "): ";
    for (int i = 0; i < nodeCount_; ++i)
        for (int j = (directed_ ? 0 : i+1); j < nodeCount_; ++j)
            if (i != j && adj_[Idx(i,j)])
                oss << i << "->" << j << "(" << weights_[Idx(i,j)] << ") ";
    return oss.str();
}

std::vector<bool> GraphGenome::FlattenAdjacency() const {
    std::vector<bool> flat;
    for (int i = 0; i < nodeCount_; ++i)
        for (int j = (directed_ ? 0 : i+1); j < nodeCount_; ++j)
            if (i != j) flat.push_back(adj_[Idx(i,j)]);
    return flat;
}

void GraphGenome::UnflattenAdjacency(const std::vector<bool>& flat) {
    size_t k = 0;
    for (int i = 0; i < nodeCount_ && k < flat.size(); ++i)
        for (int j = (directed_ ? 0 : i+1); j < nodeCount_ && k < flat.size(); ++j)
            if (i != j) SetEdge(i, j, flat[k++], weights_[Idx(i,j)]);
}

std::vector<int> GraphGenome::Neighbors(int node) const {
    std::vector<int> n;
    for (int j = 0; j < nodeCount_; ++j)
        if (j != node && adj_[Idx(node,j)]) n.push_back(j);
    return n;
}

bool GraphGenome::IsConnected() const {
    if (nodeCount_ == 0) return true;
    std::unordered_set<int> visited;
    std::queue<int> q;
    q.push(0); visited.insert(0);
    while (!q.empty()) {
        int cur = q.front(); q.pop();
        for (int nb : Neighbors(cur))
            if (visited.insert(nb).second) q.push(nb);
    }
    return (int)visited.size() == nodeCount_;
}

} // namespace NeuroSim
